### PR TITLE
Use serializable transactions in signing API task

### PR DESCRIPTION
**Pull request by mstriemer**
This adds a decorator to be used in place of `@transaction.atomic` that will first set the isolation level to `SERIALIZABLE` for the next transaction so that transactions work as you would expect. The isolation level to return to normal after the transaction ends.

### CODE REVIEW

**Comment by mstriemer** [comment]
MySQL docs: dev.mysql.com/doc/refman/5.5/en/set-transaction.html

**Comment by mstriemer** [comment]
I'll put that link in the docstring

**Comment by mstriemer** [comment]
Docstring updated.

File: apps/amo/decorators.py [hunk 1]
```diff
@@ -238,3 +239,19 @@ def wrapper(request, username, *args, **kw):
...
+def atomic(fn):
+    """Set the transaction isolation level to SERIALIZABLE and then delegate
+    to transaction.atomic to run the specified code atomically. The
+    SERIALIZABLE level will run SELECTs in LOCK IN SHARE MODE when used in
+    conjunction with transaction.atomic.
+    Docs: https://dev.mysql.com/doc/refman/5.5/en/set-transaction.html.
+    """
+    @functools.wraps(fn)
+    def inner(*args, **kwargs):
+        cursor = connection.cursor()
```
Comment (kumar303) [inline apps/amo/decorators.py]: I guess this will only work when pinned to master? Maybe it doesn't matter. I'm not sure. Hmm.

**Comment by kumar303** [comment]
I think it would be better to do this for all transactions using the init command so I'd suggest maybe putting a TODO comment in here about uplifting it to settings.

File: apps/amo/decorators.py (hunk 1 above)
Comment (mstriemer) [inline apps/amo/decorators.py]: Yeah, this should probably do the pinning.

File: apps/amo/decorators.py [hunk 2]
```diff
@@ -238,3 +239,19 @@ def wrapper(request, username, *args, **kw):
...
+    """Set the transaction isolation level to SERIALIZABLE and then delegate
+    to transaction.atomic to run the specified code atomically. The
+    SERIALIZABLE level will run SELECTs in LOCK IN SHARE MODE when used in
+    conjunction with transaction.atomic.
+    Docs: https://dev.mysql.com/doc/refman/5.6/en/set-transaction.html.
+    """
+    @functools.wraps(fn)
+    def inner(*args, **kwargs):
+        cursor = connection.cursor()
+        cursor.execute('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE')
```
Comment (kumar303) [inline apps/amo/decorators.py]: this hurts my brain a bit trying to think about master and slave connections. Is the cursor consistent (either master or slave) throughout the request? I don't even know what it would mean to start a transaction on a slave db that might rely on master changes. So perhaps this is something that should only be used when pinned to master anyway.

**Comment by kumar303** [comment]
r+wc

File: apps/amo/decorators.py (hunk 2 above)
Comment (kumar303) [inline apps/amo/decorators.py]: Is it true that this cursor change will affect all code that runs after the transaction block? I.E. all queries that run throughout the request.

Comment (mstriemer) [inline apps/amo/decorators.py]: I wrapped `inner` in `@write` so that should remove some of those edge cases.

Comment (kumar303) [inline apps/amo/decorators.py]: good idea

**Comment by kumar303** [comment]
Is it true that this cursor change will affect all code that runs after the transaction block? I.E. all queries that run throughout the request.

**Comment by mstriemer** [comment]
The docs say "Without any SESSION or GLOBAL keyword, the statement applies to the next (not started) transaction performed within the current session. Subsequent transactions revert to using the SESSION isolation level." so once the `with transaction.atomic():` block exits the transaction level should return to whatever it was before.

File: apps/amo/decorators.py [hunk 3]
```diff
@@ -238,3 +240,25 @@ def wrapper(request, username, *args, **kw):
...
+    """
+    # TODO: Make this the default for all transactions.
+    @functools.wraps(fn)
+    @write
+    def inner(*args, **kwargs):
+        cursor = connection.cursor()
+        try:
+            cursor.execute('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE')
+        except OperationalError:
+            # Looks like we're in a transaction already, too late now.
```
Comment (kumar303) [inline apps/amo/decorators.py]: can you log this error. I know it's mostly for the tests but it could theoretically happen in prod.

File: apps/amo/decorators.py [hunk 4]
```diff
@@ -238,3 +240,25 @@ def wrapper(request, username, *args, **kw):
...
+    Docs: https://dev.mysql.com/doc/refman/5.6/en/set-transaction.html.
+    """
+    # TODO: Make this the default for all transactions.
+    @functools.wraps(fn)
+    @write
+    def inner(*args, **kwargs):
+        cursor = connection.cursor()
+        try:
+            cursor.execute('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE')
+        except OperationalError:
```
Comment (kumar303) [inline apps/amo/decorators.py]: This catches TransactionManagementError too, right?

Comment (kumar303) [inline apps/amo/decorators.py]: better question: is this related? https://sentry.prod.mozaws.net/operations/olympia-stage/group/157314/ :) I'm not sure what the source of that is actually

Comment (mstriemer) [inline apps/amo/decorators.py]: That looks like it should be unrelated. Ideally this check wouldn't be needed. I might mock this decorator in the tests if that's possible.

**Comment by mstriemer** [comment]
@kumar303 Latest commit has a workaround for the test failures, what do you think?

**Comment by kumar303** [comment]
Looks like maybe you missed a couple tests but I think mocking is the right approach. You could also mock `amo.decorators.connection` because it's only used to get the cursor. Removing the create version function entirely from test coverage seems a little heavyweight.

**Comment by mstriemer** [comment]
I didn't remove the tests for `create_version_for_upload` I'm just testing it directly instead of through `submit_file` now.

**Comment by mstriemer** [comment]
Good call on mocking the cursor though, I'll try that.

**Comment by mstriemer** [comment]
So mocking the cursor made the tests not even start. I'm not sure how but I guess it was mocking the regular database connection. I updated the views to use the `non_atomic` version of `devhub.tasks.create_version_for_upload` and now they pass.

**Comment by kumar303** [comment]
r+

deal-with-it
